libs/utils/trace_analysis.py

```diff
@@ -912,196 +810,354 @@ def plotCPUFrequencyResidency(self, cpus=None):
-            plt.figlegend(handles,
-                          labels,
-                          loc='lower right')
-            plt.subplots_adjust(hspace=0.5)
+                                  figsize=(16,yrange));
+
+            axes.set_title('CPU{}'.format(cpu))
+
+            axes.grid(True);
+            if (idx+1 < n_cpus):
+                axes.set_xticklabels([])
+            else:
+                axes.set_xlabel('Time [s]')
+            axis.append(axes)
+
+        # Set same x-axes limit for all plots
+        for ax in axis:
+            ax.set_xlim(0, 1.05 * xmax)
+
+        # Put title on top of the figure. As of now there is no clean way to
+        # make the title appear always in the same position in the figure
+        # because figure heights may vary between different platforms
+        # (different number of OPPs). Hence, we use annotation
+        legend_y = axis[0].get_ylim()[1]
+        axis[0].annotate('OPP Residency Time',
+                         xy=(0, legend_y),
+                         xytext=(-50, 45),
+                         textcoords='offset points',
+                         fontsize=18)
+        axis[0].annotate('GREEN: Total',
+                         xy=(0, legend_y),
+                         xytext=(-50, 25),
+                         textcoords='offset points',
+                         color='g',
+                         fontsize=14)
+        axis[0].annotate('RED: Active',
+                         xy=(0, legend_y),
+                         xytext=(50, 25),
+                         textcoords='offset points',
+                         color='r',
+                         fontsize=14)
+        figname = '{}/{}cpu_freq_residency.png'.format(self.plotsdir,
+                                                       self.prefix)
+        pl.savefig(figname, bbox_inches='tight')
+
+    def _getClusterActiveSignal(self, cluster):
+        """
+        Build a square wave representing the active (i.e. non-idle) cluster
+        time, i.e.:
+            cluster_active[t] == 1 if at least one CPU is reported to be
+                                   non-idle by CPUFreq at time t
+            cluster_active[t] == 0 otherwise
+
+        :param cluster: list of CPU IDs belonging to a cluster
+        :type cluster: list(int)
+        """
+        idle_df = self.trace.df('cpu_idle')
+
+        cpu_active = {}
+        for cpu in cluster:
+            cpu_idle = idle_df[idle_df.cpu_id == cpu]
+
+            # time instants where CPU goes to an idle state
+            idle_times = cpu_idle[cpu_idle.state != NON_IDLE_STATE]
+            # time instants where CPU leaves an idle state
+            non_idle_times = cpu_idle[cpu_idle.state == NON_IDLE_STATE]
+
+            active_indexes = cpu_idle.index.values.tolist()
+
+            exist_zero = active_indexes[0] == 0.0
+            if not exist_zero:
+                active_indexes.insert(0, 0.0)
+
+            if idle_times.index[0] < non_idle_times.index[0]:
```

Comment:
The following block is quite "convoluted"... maybe we can simplify a bit at least to avoid repetition of the assignment to active?... I mean using a var to defined i or i+1...